test/test_biogeochemistry.jl

```diff
@@ -150,9 +153,92 @@ outputs = (w = model.velocities.w,
 simulation.output_writers[:simple_output] =
     JLD2OutputWriter(model, outputs,
                      schedule = TimeInterval(20minutes),
-                     filename = "convecting_NPD.jld2",
+                     filename = "biogeochemistry_test.jld2",
                      overwrite_existing = true)
 
+run!(simulation)
+
+#####
+##### Example using SomethingBiogeochemistry
+#####
+
+@inline growth(x, y, z, t, P, PAR, μ₀, k, m, other_params...) = (μ₀ * (1 - exp(-PAR/k)) - m) * P
+
+@inline function integrate_light!(bgc, model)
+    # have to slightly rewrite to pass parameters
+    par_calculation =  launch!(model.architecture, model.grid, :xy, update_PAR!,
+                               model.auxiliary_fields.PAR, model.grid, model.tracers.P, model.clock.time, bgc.parameters,
+                               dependencies = Event(device(model.architecture)))
+    wait(device(model.architecture), par_calculation)
+end
+
+biogeochemistry_parameters = (
+    growth_rate = 1/day,
+    light_limit = 3.5,
+    mortality_rate = 0.1/day,
+
+    water_light_attenuation_coefficient = 0.12,
+    phytoplankton_light_attenuation_coefficient = 0.06,
+    phytoplankton_light_attenuation_exponent = 0.6,
+    surface_PAR = t -> 100*max(0.0, sin(t*π/(12hours)))
+)
+
+biogeochemistry = SomethingBiogeochemistry(tracers = :P, 
+                                           auxiliary_fields = :PAR, 
```

Comment:
I don't think we want to construct auxiliary fields automagically, because this is restrictive (eg, we have to assume the fields have a certain location?)

I think we should demonstrate "proper" usage of auxiliary fields with non-trivial compute methods; ie we define `compute!` for them. This is vastly preferred especially because it will allow the abstraction to be used both for output and during post processing without boiler plate.

Reply:
Okay I've had a crack at implimenting integral fields. In case you don't think we want an abstract version of it and it should just be a part of a biogeochemical model (although I'm not sure how you could do that without a lot of what I've written for it), I've put it in a different branch [here](https://github.com/CliMA/Oceananigans.jl/tree/jsw/integral-fields). I'm also not exactly sure how we can define indexing an integral since the whole thing needs to be computed (or at least the columns being indexed).

I've also only done it for $\int^0_{-z_k}F\left[<:\text{Any}, <:\text{Any}, <:\text{Center}\right]dz$ for now because I can't see a way todo it generally so will need to repeat for the different axes/directions/field locations like diriviatives does.

<details>
  <summary>Some rambelling about why I think we have todo it like this (unlike how the Integral field reduction is done)</summary>
  I also just realised that `Integral` is a field reduction already provided giving the total integral along an axis. That is implimented as a sum of `field*volume` so I tried rethinking how I've done it to be as simple as that but don't think its correct todo just `field*volume` in this case because the integral between points at the same location as the field being integrated is an integral over half of two different cells. 
</details>

Reply:
As I've written it we could then write (for the simple light attenuaiton I implimented in the test)
```julia
PAR_fraction = Field(exp(-∫⁻dz(χ*P^e + k)))
```
Although this might mean we have to add right at the end of the model defimnition after the `tracers = TracerFields...` line?

Reply:
IntegralField is cool! I suggest we merge this PR without that though because I think it will be a fairly extensive effort to make IntegralField work properly and benchmarked and everything (it may require different methods since accumulation is fairly different on CPU vs GPU). Instead we should use this method:

```julia
struct PAROperand
    # parameters here, remember to concretely type
end

# In principle, we can restrict to Center, Center, Center, but we may not want to:
const PARField = Field{<:Any, <:Any, <:Any, <:PAROperand}

import Oceananigans.Fields: compute!

function compute!(par::PARField)
    # define here
end
```

Reply:
PS what does "PAR" stand for?

Reply:
Makses sense, after I started I realised it was a bigger challenge than I thought it would be!

I'll try and make these changes later.

PAR stands for photosynthetically available radiation, perhaps it would be more clear to call it light here.

Reply:
or how about `PhotosytheticAvailableRadiation`?

Reply:
`Light` seems pretty simple too though. Or simply `AvailableRadiation`.

Reply:
Just trying to impliment this now and the issue is that because `compute`! doesn't get the model the `P` field needs to be defined before the `PAR` field can be. Therefore we either need to manually define the tracers before, or perhaps in `validate_biogeochemistry!` make the PAR field?

Reply:
Okay, in that case we should probably just compute `P` in `update_biogeochemical_state!`.